### DISCUSSION TITLE
Issue with dual serial ports from arduino due

### DISCUSSION THREAD

**Discussion by amagraham** [bonsai-rx/bonsai#1221] [Q&A]
Hi all, 
I'm using Bonsai to both send information via serial to an arduino to control a series of outputs, as well as receive the data from the arduino to write to a CSV file.  However, I encounter an issue whenever I try to run this program on the Arduino Due, which has two USB ports-- one for sending on and one for receiving data.  I have been unable to successfully execute the protocol, even if I create two separate arduino serial connections via nodes in bonsai-- I receive an error-- "the given port name does not start with COM/com or does not resolve to a valid serial port Parameter Name portName" despite identifying these unique serial connections as different ports (e.g. COM5 and COM7) within the Bonsai nodes. 

Does anyone have any advice or idea of how to address this issue?  My arduino program also identifies and opens both serial connections.  They do work on other serial monitors, but I have not had success using bonsai. 

Thanks in advance!
![PortName Error](https://user-images.githubusercontent.com/62187791/212116558-d6a49d7f-dc65-478c-95e3-eeaac3e02e67.PNG)

Alice

**Comment by bruno-f-cruz** [comment]
Can you attach the workflow?
thanks

**Reply by amagraham** [reply]
Yes, sorry about that! I had trouble attaching but here is the xml output. 
Thanks!


![dueroutine](https://user-images.githubusercontent.com/62187791/212209675-01508b41-985b-4a35-a797-d2d1f2237e6c.PNG)

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:io="clr-namespace:Bonsai.IO;assembly=Bonsai.System"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="io:CreateSerialPort">
          <io:Name>Arduino</io:Name>
          <io:PortName>COM5</io:PortName>
          <io:BaudRate>9600</io:BaudRate>
          <io:Parity>None</io:Parity>
          <io:ParityReplace>63</io:ParityReplace>
          <io:DataBits>8</io:DataBits>
          <io:StopBits>One</io:StopBits>
          <io:Handshake>None</io:Handshake>
          <io:DiscardNull>false</io:DiscardNull>
          <io:DtrEnable>true</io:DtrEnable>
          <io:RtsEnable>false</io:RtsEnable>
          <io:ReadBufferSize>4096</io:ReadBufferSize>
          <io:WriteBufferSize>2048</io:WriteBufferSize>
          <io:ReceivedBytesThreshold>1</io:ReceivedBytesThreshold>
        </Combinator>
      </Expression>
      <Expression xsi:type="io:CsvReader">
        <io:FileName>testinput.csv</io:FileName>
        <io:ListSeparator />
        <io:ScanPattern>%s</io:ScanPattern>
        <io:SkipRows>0</io:SkipRows>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT0S</rx:DueTime>
          <rx:Period>PT15S</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Zip" />
      </Expression>
      <Expression xsi:type="MemberSelector">
        <Selector>Item1</Selector>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="io:SerialStringWrite">
          <io:PortName>Arduino</io:PortName>
          <io:NewLine>\r\n</io:NewLine>
        </Combinator>
      </Expression>
      <Expression xsi:type="Parse">
        <Pattern>&lt;%c,%i,%i&gt;</Pattern>
      </Expression>
      <Expression xsi:type="MemberSelector">
        <Selector>Item3</Selector>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="io:SerialStringRead">
          <io:PortName>ReceiveDat</io:PortName>
          <io:NewLine>\r\n</io:NewLine>
        </Combinator>
      </Expression>
      <Expression xsi:type="io:CsvWriter">
        <io:FileName>113022\testout113022_4.csv</io:FileName>
        <io:Append>false</io:Append>
        <io:Overwrite>true</io:Overwrite>
        <io:Suffix>None</io:Suffix>
        <io:IncludeHeader>true</io:IncludeHeader>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="io:CreateSerialPort">
          <io:Name>ReceiveDat</io:Name>
          <io:PortName>COM7</io:PortName>
          <io:BaudRate>9600</io:BaudRate>
          <io:Parity>None</io:Parity>
          <io:ParityReplace>63</io:ParityReplace>
          <io:DataBits>8</io:DataBits>
          <io:StopBits>One</io:StopBits>
          <io:Handshake>None</io:Handshake>
          <io:DiscardNull>false</io:DiscardNull>
          <io:DtrEnable>false</io:DtrEnable>
          <io:RtsEnable>false</io:RtsEnable>
          <io:ReadBufferSize>4096</io:ReadBufferSize>
          <io:WriteBufferSize>2048</io:WriteBufferSize>
          <io:ReceivedBytesThreshold>1</io:ReceivedBytesThreshold>
        </Combinator>
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="1" To="3" Label="Source1" />
      <Edge From="2" To="3" Label="Source2" />
      <Edge From="3" To="4" Label="Source1" />
      <Edge From="4" To="5" Label="Source1" />
      <Edge From="4" To="6" Label="Source1" />
      <Edge From="6" To="7" Label="Source1" />
      <Edge From="8" To="9" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

**Reply by bruno-f-cruz** [reply]
Can you just add the zip file with the workflow as [suggested in the announcements](https://github.com/orgs/bonsai-rx/discussions/864) post? Thanks :)

**Reply by amagraham** [reply]
[DueRoutine.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/10407304/DueRoutine.bonsai.zip)

Whoops! Must have missed it :/ thanks again!

**Reply by bruno-f-cruz** [reply]
Move the resource initialization (i.e. `CreateSerialPort`) before calling the serial port in `SerialStringWrite`. I think that should fix it.
Good luck!
